src/gluonts/mx/block/regularization.py

```diff
@@ -0,0 +1,179 @@
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+# Standard library imports
+from typing import Optional, List
+
+# Third-party imports
+from mxnet.gluon.loss import Loss
+
+# First-party imports
+from gluonts.core.component import validated
+from gluonts.model.common import Tensor
+
+
+class ActivationRegularizationLoss(Loss):
+    r"""Computes Activation Regularization Loss. (alias: AR)
+    The formulation is as below:
+    .. math::
+    L = \alpha L_2(h_t)
+    where :math:`L_2(\cdot) = {||\cdot||}_2, h_t` is the output of the RNN at timestep t.
+    :math:`\alpha` is scaling coefficient.
+    The implementation follows the work::
+    @article{merity2017revisiting,
+    title={Revisiting Activation Regularization for Language RNNs},
+    author={Merity, Stephen and McCann, Bryan and Socher, Richard},
+    journal={arXiv preprint arXiv:1708.01009},
+    year={2017}
+    }
+    Parameters
+    ----------
+    alpha : float, default 0
+        The scaling coefficient of the regularization.
+    weight : float or None
+        Global scalar weight for loss.
+    batch_axis : int, default 0
+        The axis that represents mini-batch.
+    time_axis : int, default 0
+        The axis that represents time-step.
+    """
+
+    @validated()
+    def __init__(
+        self,
+        alpha: float = 0,
+        weight: Optional[float] = None,
+        batch_axis: int = 1,
+        time_axis: int = 0,
+        **kwargs
+    ):
+        super(ActivationRegularizationLoss, self).__init__(
+            weight, batch_axis, **kwargs
+        )
+        self._alpha = alpha
+        self._batch_axis = batch_axis
+        self._time_axis = time_axis
+
+    def __repr__(self):
+        s = "ActivationRegularizationLoss (alpha={alpha})"
+        return s.format(alpha=self._alpha)
+
+    def hybrid_forward(self, F, *states: List[Tensor]) -> Tensor:
+        """
+        Parameters
+        ----------
+        states : list
+            the stack outputs from RNN, which consists of output from each time step (TNC).
+        Returns
+        --------
+        loss : NDArray
+            loss tensor with shape (batch_size,). Dimensions other than batch_axis are averaged out.
+        """
+        if self._alpha != 0:
+            if states:
+                means = []
+                for state in states:
+                    if isinstance(state, list):
+                        state = F.stack(*state, axis=self._time_axis)
+                    means.append(
+                        self._alpha
+                        * state.__pow__(2).mean(
```

Comment:
The paper and the docstring mentions that the regularization is L2 norm not squared L2 norm. But this function (also TAR) seems to compute squared L2 norm of the state vector (and averaged over batch). Am I missing something?

Reply:
it uses squared l2 norm. see official implementation: https://github.com/salesforce/awd-lstm-lm/blob/32fcb42562aeb5c7e6c9dec3f2a3baaaf68a5cb5/main.py#L201-L203

Reply:
Then update the docstring.